File: CODE/Quartic.f90
!*************************************************
!	FUNZIONE ZIRILLI2
!*************************************************
      	SUBROUTINE FUNCT(N,X,F)
	IMPLICIT DOUBLE PRECISION (A-H,O-Z)
	DIMENSION X(N)
	F=(1./4.)*X(1)**4-(1./2.)*X(1)**2+.1*X(1)+(1./2.)*X(2)**2
       	RETURN
	END
!*************************************************
!	GRADIENTE
!*************************************************
	SUBROUTINE GRAD(N,X,G)
	IMPLICIT DOUBLE PRECISION (A-H,O-Z)
	DIMENSION X(N),G(N)
	G(1)=X(1)**3-X(1)+.1
	G(2)=X(2)
	RETURN
	END
!*************************************************	
!	HESSIANO
!*************************************************
	SUBROUTINE HESS(N,X,H)
	IMPLICIT DOUBLE PRECISION (A-H,O-Z)
	DIMENSION X(N),H(N,N)
	H(1,1)=3.*X(1)**2-1.
	H(1,2)=0.
	H(2,1)=H(1,2)
	H(2,2)=1.
	RETURN
	END
!*************************************************
!	HESSIANO x DIREZIONE
!*************************************************
	SUBROUTINE HESDIR(N,X,D,HD)
	IMPLICIT DOUBLE PRECISION (A-H,O-Z)
	DIMENSION X(N),HD(N),D(N),H(2,2)
      	CALL HESS(N,X,H)
        DO 2 I=1,N
      	HD(I)=0.	 
      	DO 2 J=1,N
    2  	HD(I)=HD(I)+H(I,J)*D(J)					
     	RETURN
	END

File: Python/quartic.py
#*************************************************
#	FUNZIONE ZIRILLI2
#*************************************************

import numpy as np

def funct(x):
    f = (1./4.) * x[0]**4 - (1./2.) * x[0]**2 + 0.1 * x[0] + (1./2.) * x[1]**2
    return f

def grad(x):
    g = np.zeros(len(x))
    g[0] = x[0]**3 - x[0] + 0.1
    g[1] = x[1]
    return g

def hess(x):
    h = np.array([[3.0 * x[0]**2 - 1.0, 0.0],
                  [0.0, 1.0]])
    return h

def hesdir(x, d):
    h = hess(x)
    hd = np.dot(h, d)
    return hd

if __name__ == "__main__":

    # Example Usage
    n_value = 2
    x_values = np.array([1.0, 2.0])  # Replace with your actual input values
    direction_values = np.array([0.5, 0.8])  # Replace with your actual direction values

    result_funct = funct(x_values)
    result_grad = grad(x_values)
    result_hess = hess(x_values)
    result_hesdir = hesdir(x_values, direction_values)

    print("Result (Function):", result_funct)
    print("Result (Gradient):", result_grad)
    print("Result (Hessian):", result_hess)
    print("Result (Hessian x Direction):", result_hesdir)
